### PR TITLE
Review

### PR DESCRIPTION
**Pull request by Tawakalt**

### CODE REVIEW

File: places/views.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+from django.http import JsonResponse
+import requests, json, os
+from dotenv import load_dotenv
+
+# reading .env file
+load_dotenv()
```
Comment (abiodun0) [inline places/views.py]: Why do you have to call load_dotenv() here again when you already called in settings.py?

File: places/views.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+
+def index(request): 
+    q = request.GET.get("q", False)
```
Comment (abiodun0) [inline places/views.py:7]: There is a lot to unpack here, you can either search for places or lat and long, your implementation here is saying you can search for both


Even if that's the case, you didn't take care of when they send empty location

So this function should have been like this

```
       q = request.GET.get("q", '')
        lng = request.GET.get("lng", '')
        lat = request.GET.get("lat", '')
        API_KEY = os.getenv('API_KEY')
        searchUrl = "https://maps.googleapis.com/maps/api/place/findplacefromtext/json?inputtype=textquery&"
         searchResponse = requests.get(f'{searchUrl}input={q}&locationbias=point:{lng},{lat}&key={API_KEY}') 
        ```

And there shouldn't be anything wrong with it

File: places/views.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                'Name': result['name'],
+                'Location': {
+                    'Lat': result['geometry']['location']['lat'],
+                    'Lng': result['geometry']['location']['lng'],
+                },
+                'Address': result['formatted_address'],
+                'URI': result['url']
+            }
+            status = 'OK'
+        else:
```
Comment (abiodun0) [inline places/views.py]: This else is not required

You could apply returning early here and avoid nested ifs

File: search/settings.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/3.0/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = os.getenv('SECRET_KEY')
+
+# SECURITY WARNING: don't run with debug turned on in production!
+DEBUG = os.getenv('DJANGO_DEBUG')
+
+ALLOWED_HOSTS = ['localhost', '127.0.0.1', 'place-search.herokuapp.com']
```
Comment (abiodun0) [inline search/settings.py:32]: This should be among your env variables so that it doesn't get exposed

File: search/settings.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+
+# Application definition
+
+INSTALLED_APPS = [
+    'django.contrib.admin',
```
Comment (abiodun0) [inline search/settings.py]: There is a lot of apps youdon't need here, you didn't use the auths, nor sessions, nor messages, perhaps remove them?

File: places/urls.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+from django.urls import path
+from . import views
+
+urlpatterns = [
```
Comment (abiodun0) [inline places/urls.py:4]: you mentioned that you need it to be extendable, right here is where you can put the keyward param, here you can have {placeApi}
so in the index, you check for param and if it's not google, you can return `This place api is still in development`

**Review by abiodun0 — CHANGES_REQUESTED** [review]
Looking into this again, there are a couple of things to note

1. You only have one endpoint, which means this could easily have been done in flask without any ORM, but Django has ORM inbuilt, so maybe overkill?

2. You didn't mock most of the external endpoint server during tests to return the required results

File: places/views.py (hunk 2 above)
Comment (Tawakalt) [reply places/views.py:7]: Google API expects places. You can decide to include the lat and long for more specific result. It's not either places or lat/lng

I handled empty location

Did you just change `False` to `''`?

File: places/views.py (hunk 1 above)
Comment (Tawakalt) [reply places/views.py]: I'll take it out

File: places/views.py (hunk 3 above)
Comment (Tawakalt) [reply places/views.py]: Ok

File: search/settings.py (hunk 4 above)
Comment (Tawakalt) [reply search/settings.py:32]: But the deployed app is not private na!
The url is on my repo description already

File: search/settings.py (hunk 5 above)
Comment (Tawakalt) [reply search/settings.py]: Yeah, I'll take them out.

File: places/urls.py (hunk 6 above)
Comment (Tawakalt) [reply places/urls.py:4]: Hmm....ok

File: places/views.py (hunk 2 above)
Comment (abiodun0) [reply places/views.py:7]: Yes, they are both falsy values

Comment (Tawakalt) [reply places/views.py:7]: So either is good?
Or is `''` better?

I'm tryna understand the reason behind the change ni o

Comment (abiodun0) [reply places/views.py:7]: so you don't need to use an if anywhere, `''` already satisfies that if it's not required its a wild card search

Comment (Tawakalt) [reply places/views.py:7]: But I need the if to handle empty or False search

File: places/sources.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    q = request.GET.get("q", False)
+    lng = request.GET.get("lng", False)
+    lat = request.GET.get("lat", False)
+    API_KEY = os.getenv('API_KEY')
+    searchUrl = "https://maps.googleapis.com/maps/api/place/findplacefromtext/json?inputtype=textquery&"
+
+    if lng and lat:
+        searchResponse = requests.get(f'{searchUrl}input={q}&locationbias=point:{lng},{lat}&key={API_KEY}') 
+    else:
+        searchResponse = requests.get(f'{searchUrl}input={q}&key={API_KEY}')
```
Comment (abiodun0) [inline places/sources.py]: You don't need the else, put this at the top of the if 
```
searchResponse = requests.get(f'{searchUrl}input={q}&key={API_KEY}')
```

File: places/sources.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
+from django.http import JsonResponse
+import requests, json, os
+
+
+def google(request):
+    q = request.GET.get("q", False)
+    lng = request.GET.get("lng", False)
```
Comment (abiodun0) [inline places/sources.py]: Since this is.a string, do it like this instead `lng = request.GET.get("lng", '')` same for lat

File: places/views.py [hunk 9]
```diff
@@ -0,0 +1,29 @@
+from django.http import JsonResponse
+import requests, json, os
+from . import sources
```
Comment (abiodun0) [inline places/views.py]: This would have been more explicit if you do 

`from .source import google`

File: places/tests.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        "Location": {"Lat": LAT, "Lng": LNG}, 
+        "Address": "Lagos, Nigeria", 
+        "URI": "https://maps.google.com/?q=Lagos,+Nigeria&ftid=0x103b8b2ae68280c1:0xdc9e87a367c3d9cb"
+    }
+}
+ZERO_RESULT = {'status': 'ZERO_RESULTS', 'result': []}
+
+
+# Create your tests here.
+class SearchTest(TestCase):
```
Comment (abiodun0) [inline places/tests.py:32]: Shouldn't there be something like a setUp method here?

**Review by abiodun0 — COMMENTED** [review]
Looks good, just a few nitpicks

File: places/sources.py (hunk 7 above)
Comment (Tawakalt) [reply places/sources.py]: Doing this breaks my test weirdly

File: places/tests.py (hunk 10 above)
Comment (Tawakalt) [reply places/tests.py:32]: Not needed

File: places/tests.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+LNG = 3.3792057
+LAT = 6.5243793
+PLACE_ID = "ChIJwYCC5iqLOxARy9nDZ6OHntw"
+COMMON = "https://maps.googleapis.com/maps/api/place/"
+URL = f"{COMMON}findplacefromtext/json?inputtype=textquery&key={API_KEY}&input="
+ZERO_SEARCH_URL = f"{URL}xxxx"
+SEARCH_URL = f"{URL}{Q}"
+SEARCH_URL_WITH_LOCATION = f"{SEARCH_URL}&locationbias=point:{LNG},{LAT}"
+DETAILS_URL = f"{COMMON}details/json?place_id={PLACE_ID}&key={API_KEY}"
+DETAILS = mock.details
```
Comment (abiodun0) [inline places/tests.py]: be explicit about what you are importing
do 
`from .mock import no_candidate, candidate... ` instead

**Pull request closed without merging**